apollo-router/src/plugins/file_uploads/multipart_form_data.rs

```diff
@@ -0,0 +1,89 @@
+use bytes::Bytes;
+use bytes::BytesMut;
+use futures::stream::StreamExt;
+use futures::stream::TryStreamExt;
+use futures::Stream;
+use http::HeaderMap;
+use http::HeaderValue;
+use mediatype::names::BOUNDARY;
+use mediatype::names::FORM_DATA;
+use mediatype::names::MULTIPART;
+use mediatype::MediaType;
+use rand::RngCore;
+
+use super::MapField;
+use super::MultipartRequest;
+use super::UploadResult;
+
+pub(super) struct MultipartFormData {
+    boundary: String,
+    operations: hyper::Body,
+    map: MapField,
+    multipart: MultipartRequest,
+}
+
+impl MultipartFormData {
+    pub(super) fn new(operations: hyper::Body, map: MapField, multipart: MultipartRequest) -> Self {
+        let boundary = format!(
+            "------------------------{:016x}",
+            rand::thread_rng().next_u64()
+        );
+        Self {
+            boundary,
+            operations,
+            map,
+            multipart,
+        }
+    }
+
+    pub(super) fn content_type(&self) -> HeaderValue {
+        let boundary =
+            mediatype::Value::new(&self.boundary).expect("boundary should be valid value");
+        let params = [(BOUNDARY, boundary)];
+        let mime = MediaType::from_parts(MULTIPART, FORM_DATA, None, &params);
+        mime.to_string()
+            .try_into()
+            .expect("mime should be valid header value")
+    }
+
+    pub(super) async fn into_stream(mut self) -> impl Stream<Item = UploadResult<Bytes>> {
+        let map_bytes = serde_json::to_vec(&self.map).expect("map should be serializable to JSON");
+        let field_prefix = |name: &str| {
+            tokio_stream::once(Ok(Bytes::from(format!(
```

Comment:
generally we use `once` from the `futures` crate instead of reaching for a transitive dependency https://docs.rs/futures/latest/futures/stream/fn.once.html